File: src/drivers/arduino_driver/scripts/arduino_bridge_node.py
#!/usr/bin/env python3
"""Arduino serial bridge: command output + ultrasonic input.

Arduino 명령어:
  - F: 전진 + 직진
  - B: 후진 + 직진
  - L: 전진 + 최대 좌회전
  - R: 전진 + 최대 우회전
  - l: 전진 + 약한 좌회전
  - r: 전진 + 약한 우회전
  - S: 정지
"""

from __future__ import annotations

import math
import threading
from typing import Dict

import rclpy
from rclpy.node import Node
from ackermann_msgs.msg import AckermannDrive
from std_msgs.msg import Float32MultiArray

try:
    import serial
except ImportError as exc:  # pragma: no cover
    raise RuntimeError("pyserial is required for arduino_bridge_node") from exc


SENSOR_ORDER = ("F", "FL", "FR", "R", "RL", "RR")

def clamp(value: float, lo: float, hi: float) -> float:
    return lo if value < lo else hi if value > hi else value


def parse_ultrasonic_line(line: str) -> Dict[str, float]:
    """Parse 'F:25,FL:30,...' into meters."""
    data: Dict[str, float] = {}
    for part in line.split(","):
        if ":" not in part:
            continue
        key, value = part.split(":", 1)
        key = key.strip()
        try:
            cm = float(value.strip())
        except ValueError:
            continue
        data[key] = cm / 100.0
    return data


class ArduinoBridgeNode(Node):
    def __init__(self) -> None:
        super().__init__("arduino_bridge_node")

        self.declare_parameter("port", "/dev/ttyACM0")
        self.declare_parameter("baudrate", 115200)
        self.declare_parameter("command_topic", "/arduino/cmd")
        self.declare_parameter("ultrasonic_topic", "/ultrasonic/ranges")
        self.declare_parameter("max_speed_mps", 2.0)
        self.declare_parameter("max_steer_deg", 30.0)
        self.declare_parameter("center_servo_deg", 90.0)

        self.port = self.get_parameter("port").value
        self.baudrate = int(self.get_parameter("baudrate").value)
        self.command_topic = self.get_parameter("command_topic").value
        self.ultrasonic_topic = self.get_parameter("ultrasonic_topic").value
        self.max_speed_mps = float(self.get_parameter("max_speed_mps").value)
        self.max_steer_deg = float(self.get_parameter("max_steer_deg").value)
        self.center_servo_deg = float(self.get_parameter("center_servo_deg").value)

        # 시리얼 포트 열기 (DTR 토글로 Arduino 리셋 발생)
        self.serial = serial.Serial(self.port, self.baudrate, timeout=0.01)
        self.serial_lock = threading.Lock()

        # Arduino 리셋 후 부팅 대기 (2초)
        self.get_logger().info("[arduino_bridge] Arduino 리셋 대기 중 (2초)...")
        import time
        time.sleep(2)
        # 부팅 메시지 비우기
        if self.serial.in_waiting:
            self.serial.read(self.serial.in_waiting)
        self.get_logger().info("[arduino_bridge] Arduino 준비 완료")

        self.ultra_pub = self.create_publisher(Float32MultiArray, self.ultrasonic_topic, 1)
        self.create_subscription(AckermannDrive, self.command_topic, self.cmd_cb, 1)

        # 마지막 전송 명령 (중복 전송 방지)
        self.last_cmd = ""
        self.buffer = ""

        self.read_timer = self.create_timer(0.01, self.read_serial)

        self.get_logger().info(
            f"[arduino_bridge] port={self.port} baud={self.baudrate} mode=continuous"
        )

    def cleanup(self) -> None:
        """Send stop command to Arduino on shutdown."""
        try:
            if self.serial and self.serial.is_open:
                self.get_logger().info("[arduino_bridge] Sending stop command...")
                stop_cmd = "V:0,S:90\n"
                with self.serial_lock:
                    self.serial.write(stop_cmd.encode("utf-8"))
                    self.serial.flush()
                import time
                time.sleep(0.1)  # 명령 전송 대기
                self.get_logger().info("[arduino_bridge] Motor stopped")
        except Exception as e:
            self.get_logger().error(f"[arduino_bridge] Cleanup error: {e}")

    def cmd_cb(self, msg: AckermannDrive) -> None:
        pwm = self._speed_to_pwm(msg.speed)
        servo = self._steer_to_servo(msg.steering_angle)
        payload = f"V:{pwm},S:{servo}\n"
        with self.serial_lock:
            self.serial.write(payload.encode("utf-8"))

    def _speed_to_pwm(self, speed: float) -> int:
        # Firmware expects 0~255 PWM (forward only)
        speed = clamp(speed, 0.0, self.max_speed_mps)
        ratio = speed / max(self.max_speed_mps, 1e-3)
        return int(round(clamp(ratio * 255.0, 0.0, 255.0)))

    def _steer_to_servo(self, steering_angle: float) -> int:
        steer_deg = math.degrees(steering_angle)
        steer_deg = clamp(steer_deg, -self.max_steer_deg, self.max_steer_deg)
        servo = self.center_servo_deg + steer_deg
        return int(round(clamp(servo, self.center_servo_deg - self.max_steer_deg,
                               self.center_servo_deg + self.max_steer_deg)))

    def read_serial(self) -> None:
        try:
            waiting = self.serial.in_waiting
            if waiting <= 0:
                return
            with self.serial_lock:
                data = self.serial.read(waiting)
        except serial.SerialException as exc:
            self.get_logger().error(f"[arduino_bridge] serial error: {exc}")
            return

        if not data:
            return

        try:
            self.buffer += data.decode("utf-8", errors="ignore")
        except Exception:
            self.buffer = ""

        # 버퍼 오버플로우 방지
        if len(self.buffer) > 1024:
            self.buffer = self.buffer[-512:]

        while "\n" in self.buffer:
            line, self.buffer = self.buffer.split("\n", 1)
            parsed = parse_ultrasonic_line(line)
            if parsed:
                msg = Float32MultiArray()
                msg.data = [parsed.get(key, 0.0) for key in SENSOR_ORDER]
                self.ultra_pub.publish(msg)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = ArduinoBridgeNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        try:
            # 모터 정지 명령 전송 (CRITICAL!)
            node.cleanup()
            # 시리얼 포트 닫기
            if node.serial and node.serial.is_open:
                node.serial.close()
        except Exception:
            pass
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/drivers/arduino_driver/src/arduino_bridge_node.cpp
#include "arduino_driver/arduino_bridge_node.hpp"
#include "arduino_driver/serial_protocol.hpp"
#include <sstream>

namespace arduino_driver {

ArduinoBridgeNode::ArduinoBridgeNode()
    : Node("arduino_bridge_node"), running_(true) {
    // Declare and load parameters
    this->declare_parameter("port", "/dev/ttyACM0");
    this->declare_parameter("baudrate", 9600);
    this->declare_parameter("command_topic", "/arduino/cmd");
    this->declare_parameter("ultrasonic_topic", "/ultrasonic/ranges");
    this->declare_parameter("use_legacy_cmd", true);
    this->declare_parameter("max_speed_mps", 2.0);
    this->declare_parameter("max_steer_deg", 30.0);
    this->declare_parameter("center_servo_deg", 90.0);

    port_ = this->get_parameter("port").as_string();
    baudrate_ = this->get_parameter("baudrate").as_int();
    command_topic_ = this->get_parameter("command_topic").as_string();
    ultrasonic_topic_ = this->get_parameter("ultrasonic_topic").as_string();
    use_legacy_cmd_ = this->get_parameter("use_legacy_cmd").as_bool();
    max_speed_mps_ = this->get_parameter("max_speed_mps").as_double();
    max_steer_deg_ = this->get_parameter("max_steer_deg").as_double();
    center_servo_deg_ = this->get_parameter("center_servo_deg").as_double();

    // Setup ROS interfaces
    cmd_sub_ = this->create_subscription<ackermann_msgs::msg::AckermannDrive>(
        command_topic_, 10,
        std::bind(&ArduinoBridgeNode::cmdCallback, this, std::placeholders::_1));
    
    ultra_pub_ = this->create_publisher<std_msgs::msg::Float32MultiArray>(
        ultrasonic_topic_, 10);

    // Initialize serial port
    initSerial();

    // Start read thread
    read_thread_ = std::thread(&ArduinoBridgeNode::readLoop, this);

    RCLCPP_INFO(this->get_logger(), "ArduinoBridgeNode initialized");
    RCLCPP_INFO(this->get_logger(), "  port: %s", port_.c_str());
    RCLCPP_INFO(this->get_logger(), "  baudrate: %d", baudrate_);
    RCLCPP_INFO(this->get_logger(), "  use_legacy_cmd: %s", use_legacy_cmd_ ? "true" : "false");
}

ArduinoBridgeNode::~ArduinoBridgeNode() {
    running_ = false;

    if (serial_port_ && serial_port_->is_open()) {
        serial_port_->close();
    }

    if (read_thread_.joinable()) {
        read_thread_.join();
    }
}

void ArduinoBridgeNode::initSerial() {
    try {
        serial_port_ = std::make_unique<boost::asio::serial_port>(io_service_);
        serial_port_->open(port_);
        serial_port_->set_option(boost::asio::serial_port_base::baud_rate(baudrate_));
        serial_port_->set_option(boost::asio::serial_port_base::character_size(8));
        serial_port_->set_option(boost::asio::serial_port_base::parity(
            boost::asio::serial_port_base::parity::none));
        serial_port_->set_option(boost::asio::serial_port_base::stop_bits(
            boost::asio::serial_port_base::stop_bits::one));

        RCLCPP_INFO(this->get_logger(), "Serial port opened: %s at %d baud", 
                    port_.c_str(), baudrate_);
    } catch (const boost::system::system_error& e) {
        RCLCPP_ERROR(this->get_logger(), "Failed to open serial port %s: %s", 
                     port_.c_str(), e.what());
        throw;
    }
}

void ArduinoBridgeNode::cmdCallback(const ackermann_msgs::msg::AckermannDrive::SharedPtr msg) {
    std::lock_guard<std::mutex> lock(write_mutex_);

    if (!serial_port_ || !serial_port_->is_open()) {
        return;
    }

    try {
        if (use_legacy_cmd_) {
            std::string cmd = toLegacyCommand(*msg);
            if (cmd != last_cmd_) {
                std::string payload = cmd + "\n";
                boost::asio::write(*serial_port_, boost::asio::buffer(payload));
                last_cmd_ = cmd;
            }
        } else {
            int pwm = speedToPwm(msg->speed);
            int servo = steerToServo(msg->steering_angle);
            std::ostringstream oss;
            oss << "V:" << pwm << ",S:" << servo << "\n";
            boost::asio::write(*serial_port_, boost::asio::buffer(oss.str()));
        }
    } catch (const boost::system::system_error& e) {
        RCLCPP_ERROR_THROTTLE(this->get_logger(), *this->get_clock(), 1000,
                              "Serial write error: %s", e.what());
    }
}

std::string ArduinoBridgeNode::toLegacyCommand(const ackermann_msgs::msg::AckermannDrive& msg) {
    // Check for stop
    if (std::abs(msg.speed) < 1e-3) {
        return "S";
    }

    // Convert steering angle to degrees
    double steer_deg = radToDeg(msg.steering_angle);

    // Check for reverse
    if (msg.speed < 0) {
        return "B";
    }

    // Forward with steering
    if (steer_deg > STEER_THRESHOLD_HARD) {
        return "R";  // Hard right
    }
    if (steer_deg < -STEER_THRESHOLD_HARD) {
        return "L";  // Hard left
    }
    if (steer_deg > STEER_THRESHOLD_SOFT) {
        return "r";  // Soft right
    }
    if (steer_deg < -STEER_THRESHOLD_SOFT) {
        return "l";  // Soft left
    }

    return "F";  // Forward straight
}

int ArduinoBridgeNode::speedToPwm(double speed) {
    double clamped = clamp(std::abs(speed), 0.0, max_speed_mps_);
    double ratio = clamped / max_speed_mps_;
    return static_cast<int>(clamp(ratio * 255.0, 0.0, 255.0));
}

int ArduinoBridgeNode::steerToServo(double steering_angle) {
    double steer_deg = radToDeg(steering_angle);
    steer_deg = clamp(steer_deg, -max_steer_deg_, max_steer_deg_);
    return static_cast<int>(center_servo_deg_ + steer_deg);
}

void ArduinoBridgeNode::readLoop() {
    std::string buffer;
    char c;

    while (running_ && rclcpp::ok()) {
        try {
            if (!serial_port_ || !serial_port_->is_open()) {
                std::this_thread::sleep_for(std::chrono::milliseconds(100));
                continue;
            }

            boost::asio::read(*serial_port_, boost::asio::buffer(&c, 1));

            if (c == '\n') {
                auto parsed = parseUltrasonicLine(buffer);
                if (!parsed.empty()) {
                    std_msgs::msg::Float32MultiArray msg;
                    msg.data.reserve(SENSOR_ORDER.size());

                    for (const auto& key : SENSOR_ORDER) {
                        auto it = parsed.find(key);
                        msg.data.push_back(it != parsed.end() ? static_cast<float>(it->second) : 0.0f);
                    }

                    ultra_pub_->publish(msg);
                }
                buffer.clear();
            } else if (c != '\r') {
                buffer += c;
                // Prevent buffer overflow
                if (buffer.size() > 1024) {
                    buffer = buffer.substr(512);
                }
            }
        } catch (const boost::system::system_error& e) {
            if (running_) {
                RCLCPP_ERROR_THROTTLE(this->get_logger(), *this->get_clock(), 1000,
                                      "Serial read error: %s", e.what());
            }
        }
    }
}

std::map<std::string, double> ArduinoBridgeNode::parseUltrasonicLine(const std::string& line) {
    return parseUltrasonicData(line);
}

}  // namespace arduino_driver

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);

    try {
        auto node = std::make_shared<arduino_driver::ArduinoBridgeNode>();
        rclcpp::spin(node);
    } catch (const std::exception& e) {
        RCLCPP_ERROR(rclcpp::get_logger("arduino_bridge_node"), 
                     "Exception in arduino_bridge_node: %s", e.what());
        rclcpp::shutdown();
        return 1;
    }

    rclcpp::shutdown();
    return 0;
}
